File: packages/agentenkit/test/thread-record.test.ts
import { describe, expect, it } from 'bun:test';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { z } from 'zod';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryRunStreams, MemoryStorage } from '../src/adapters/memory.js';
import { openSqlite, SqliteStorage } from '../src/adapters/sqlite.js';
import { agentTool } from '../src/core/tools.js';
import { RECORD_EVENT_TYPES } from '../src/core/publish.js';
import { resolveConfig } from '../src/core/types.js';

// Workstream S5: the thread record keeps only what must outlive a run. The
// Go package runs the same cases under the same names (thread_record_test.go).

function scriptedModel(steps: Array<{ text?: string; call?: { id: string; name: string } }>) {
  let n = 0;
  return new MockLanguageModelV1({
    provider: 'mock',
    modelId: 'mock',
    doStream: async () => {
      const step = steps[Math.min(n++, steps.length - 1)]!;
      const chunks: LanguageModelV1StreamPart[] = [];
      if (step.text) chunks.push({ type: 'text-delta', textDelta: step.text });
      if (step.call) {
        chunks.push({ type: 'tool-call', toolCallType: 'function', toolCallId: step.call.id, toolName: step.call.name, args: '{}' });
      }
      chunks.push({ type: 'finish', finishReason: step.call ? 'tool-calls' : 'stop', usage: { promptTokens: 10, completionTokens: 5 } });
      return { stream: simulateReadableStream({ chunks }), rawCall: { rawPrompt: null, rawSettings: {} } };
    },
  });
}

async function makeRuntime(model: any) {
  const storage = new MemoryStorage();
  const queue = new MemoryQueue();
  const streams = new MemoryRunStreams();
  const runtime = await setupAgentCore({
    storage, admin: new MemoryAdminStore(), bus: new MemoryBus(), queue, kv: new MemoryKv(), streams,
    resolveModel: () => ({ instance: () => model, contextWindow: 128_000 }),
    config: resolveConfig({ streamFlushMs: 0 }),
  });
  return { runtime, storage, queue, streams };
}

describe('thread record', () => {
  it('a run writes no CHUNK to the thread record', async () => {
    const r = await makeRuntime(scriptedModel([{ text: 'looking', call: { id: 'c1', name: 'look' } }, { text: 'done' }]));
    const chat = r.runtime.createStreamTextAgent({
      name: 'chat', model: 'gpt-4o',
      tools: { look: agentTool({ parameters: z.object({}), execute: async () => 'seen' }) },
    });
    const ran = await chat.run({ prompt: 'hi' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const types = (await r.storage.events.listSince(ran.threadId, -1)).map((e) => e.type);
    expect(types).toEqual(['RUN_STARTED', 'RUN_ENDED']);
    expect(types.every((t) => RECORD_EVENT_TYPES.has(t))).toBe(true);
  });

  it('each segment starts and ends in the record', async () => {
    const r = await makeRuntime(scriptedModel([{ text: 'hi' }]));
    const chat = r.runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
    const ran = await chat.run({ prompt: 'hi' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const [started, ended] = await r.storage.events.list(ran.threadId, { types: ['RUN_STARTED', 'RUN_ENDED'] });
    expect(started).toMatchObject({ runId: ran.runId, payload: { runId: ran.runId, streamId: `${ran.runId}:1`, segment: 1 } });
    expect(ended).toMatchObject({ runId: ran.runId, payload: { streamId: `${ran.runId}:1`, status: 'finished' } });
    expect(ended!.seq).toBeGreaterThan(started!.seq);
  });

  const stores: Array<readonly [string, () => Promise<unknown>]> = [
    ['memory', async () => new MemoryStorage()],
    ['sqlite', async () => new SqliteStorage(await openSqlite(':memory:'))],
  ];
  // Prisma against a real database, through the example app's client.
  if (process.env.TEST_PRISMA_DB) {
    stores.push(['prisma', async () => {
      const { createRequire } = await import('node:module');
      const { join } = await import('node:path');
      const require = createRequire(join(import.meta.dir, '../../../examples/nextjs-app/package.json'));
      const { PrismaClient } = require('@prisma/client');
      const { PrismaStorage } = await import('../src/adapters/prisma.js');
      return new PrismaStorage(new PrismaClient({ datasources: { db: { url: process.env.TEST_PRISMA_DB } } }));
    }]);
  }
  for (const [name, open] of stores) {
    it(`the store mints seqs in order, one per entry (${name})`, async () => {
      const storage: any = await open();
      const th = await storage.threads.create({});
      const seqs = await Promise.all(
        Array.from({ length: 5 }, (_, i) =>
          storage.events.append(th.id, { type: 'CONTEXT_COMPACTED', payload: { i }, runId: i % 2 ? 'r1' : null })),
      );
      expect(seqs.map((e: any) => e.seq).sort()).toEqual([1, 2, 3, 4, 5]);
      expect(await storage.events.list(th.id, { after: 3 })).toHaveLength(2);
      expect(await storage.events.list(th.id, { limit: 2 })).toHaveLength(2);
      expect(await storage.events.list(th.id, { runId: 'r1' })).toHaveLength(2);
      expect(await storage.events.list(th.id, { types: ['INPUT_REQUIRED'] })).toHaveLength(0);
    });
  }

  it('the snapshot carries the latest run stream, not every run', async () => {
    const r = await makeRuntime(scriptedModel([{ text: 'one' }, { text: 'two' }, { text: 'three' }]));
    const chat = r.runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
    let ran: any;
    for (const prompt of ['a', 'b', 'c']) {
      ran = await chat.run({ threadId: ran?.threadId, prompt });
      await r.runtime.worker.handleJob(r.queue.items.shift()!);
    }
    const snap = await r.runtime.getThreadSnapshot(ran.threadId);
    expect(snap!.stream).toMatchObject({ streamId: `${ran.runId}:1`, runId: ran.runId, end: { status: 'finished' } });
    // The finished step is in the messages, so none of its text is in the items.
    expect(snap!.stream!.items.some((i) => i.type === 'TEXT_MESSAGE_CONTENT')).toBe(false);
    expect(snap!.stream!.offset).not.toBeNull();
  });

  it('a durable CUSTOM goes to the thread record; a plain one to the stream', async () => {
    const r = await makeRuntime(scriptedModel([{ call: { id: 'c1', name: 'go' } }, { text: 'done' }]));
    const chat = r.runtime.createStreamTextAgent({
      name: 'chat', model: 'gpt-4o',
      tools: {
        go: agentTool({
          parameters: z.object({}),
          execute: async (_a, { publishEvent }) => {
            await publishEvent('INVOICE_CREATED', { id: 'inv_1' }, { durable: true });
            await publishEvent('PROGRESS', { pct: 50 });
            return 'ok';
          },
        }),
      },
    });
    const ran = await chat.run({ prompt: 'hi' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const record = (await r.storage.events.listSince(ran.threadId, -1)).map((e) => e.type);
    expect(record).toContain('INVOICE_CREATED');
    expect(record).not.toContain('PROGRESS');
    // Each reaches a tab once: the durable one as its record entry, the
    // plain one on the stream.
    const custom = (await r.streams.snapshot(`${ran.runId}:1`))!.items
      .filter((i) => i.type === 'CUSTOM').map((i: any) => i.name);
    expect(custom).toEqual(['PROGRESS']);
  });
});

File: packages/go-agentenkit/thread_record_test.go
package agentenkit_test

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"testing"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/memory"
	pgstorage "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/postgres"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/sqlite"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Workstream S5: the thread record keeps only what must outlive a run. The
// TS package runs the same cases under the same names
// (test/thread-record.test.ts).

func TestThreadRecord(t *testing.T) {
	t.Run("a run writes no CHUNK to the thread record", func(t *testing.T) {
		h, _ := streamRuntime(t, scripted(step{text: "looking", calls: []call{{"c1", "look", `{}`}}}, step{text: "done"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{
			Name: "chat", Model: "gpt-4o",
			Tools: []agentenkit.Tool{tool("look", func(context.Context, map[string]any) (string, error) { return "seen", nil })},
		})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
		h.handleNext(t)
		stored, _ := h.storage.Events().ListSince(h.ctx, ran.ThreadID, -1, agentenkit.StorageContext{})
		var types []string
		for _, e := range stored {
			types = append(types, e.Type)
			mustEqual(t, core.RecordEventTypes[e.Type], true, e.Type+" is a record type")
		}
		mustStrings(t, types, []string{"RUN_STARTED", "RUN_ENDED"}, "the record")
	})

	t.Run("each segment starts and ends in the record", func(t *testing.T) {
		h, _ := streamRuntime(t, scripted(step{text: "hi"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
		h.handleNext(t)
		got, _ := h.storage.Events().List(h.ctx, ran.ThreadID, ports.ThreadEventFilter{Types: []string{"RUN_STARTED", "RUN_ENDED"}}, agentenkit.StorageContext{})
		mustEqual(t, len(got), 2, "two entries")
		started, ended := got[0], got[1]
		mustEqual(t, started.RunID+" "+ended.RunID, ran.RunID+" "+ran.RunID, "both belong to the run")
		mustEqual(t, payload(started)["streamId"], ran.RunID+":1", "the stream")
		mustEqual(t, payload(started)["segment"], float64(1), "segment")
		mustEqual(t, payload(ended)["streamId"], ran.RunID+":1", "the same stream")
		mustEqual(t, payload(ended)["status"], "finished", "status")
		mustEqual(t, ended.Seq > started.Seq, true, "in order")
	})

	stores := []struct {
		name string
		open func(t *testing.T) ports.Storage
	}{
		{"memory", func(*testing.T) ports.Storage { return memory.NewStorage() }},
		{"sqlite", func(t *testing.T) ports.Storage {
			db, err := sqlite.Open(filepath.Join(t.TempDir(), "record.sqlite"))
			must(t, err)
			t.Cleanup(func() { db.Close() })
			s, err := sqlite.New(db)
			must(t, err)
			return s
		}},
	}
	if os.Getenv("TEST_ADMIN_PG") != "" {
		stores = append(stores, struct {
			name string
			open func(t *testing.T) ports.Storage
		}{"postgres", func(t *testing.T) ports.Storage {
			db := openPostgres(t)
			t.Cleanup(func() { db.Close() })
			s, err := pgstorage.New(context.Background(), db)
			must(t, err)
			return s
		}})
	}
	for _, st := range stores {
		t.Run("the store mints seqs in order, one per entry ("+st.name+")", func(t *testing.T) {
			s := st.open(t)
			ctx, sc := context.Background(), ports.StorageContext{}
			th, err := s.Threads().Create(ctx, ports.ThreadInit{}, sc)
			must(t, err)
			var wg sync.WaitGroup
			seqs := make([]int64, 5)
			for i := range 5 {
				wg.Add(1)
				go func() {
					defer wg.Done()
					runID := ""
					if i%2 == 1 {
						runID = "r1"
					}
					e, err := s.Events().Append(ctx, th.ID, ports.NewThreadEvent{
						Type: "CONTEXT_COMPACTED", Payload: json.RawMessage(`{}`), RunID: runID,
					}, sc)
					if err != nil {
						t.Error(err)
					}
					seqs[i] = e.Seq
				}()
			}
			wg.Wait()
			sort.Slice(seqs, func(i, j int) bool { return seqs[i] < seqs[j] })
			mustEqual(t, fmt.Sprint(seqs), "[1 2 3 4 5]", "seqs")
			three := int64(3)
			count := func(f ports.ThreadEventFilter) int {
				got, err := s.Events().List(ctx, th.ID, f, sc)
				must(t, err)
				return len(got)
			}
			mustEqual(t, count(ports.ThreadEventFilter{After: &three}), 2, "after")
			mustEqual(t, count(ports.ThreadEventFilter{Limit: 2}), 2, "limit")
			mustEqual(t, count(ports.ThreadEventFilter{RunID: "r1"}), 2, "by run")
			mustEqual(t, count(ports.ThreadEventFilter{Types: []string{"INPUT_REQUIRED"}}), 0, "by type")
		})
	}

	t.Run("the snapshot carries the latest run stream, not every run", func(t *testing.T) {
		h, _ := streamRuntime(t, scripted(step{text: "one"}, step{text: "two"}, step{text: "three"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
		var ran agentenkit.RunResult
		for _, prompt := range []string{"a", "b", "c"} {
			ran = h.run(t, chat, agentenkit.RunInput{ThreadID: ran.ThreadID, Prompt: prompt})
			h.handleNext(t)
		}
		snap, err := h.rt.GetThreadSnapshot(h.ctx, ran.ThreadID, nil)
		must(t, err)
		mustEqual(t, snap.Stream.StreamID, ran.RunID+":1", "the latest run's stream")
		mustEqual(t, snap.Stream.RunID, ran.RunID, "its run")
		mustEqual(t, snap.Stream.End.(*ports.RunFinishedEvent).Status, "finished", "it ended")
		for _, i := range snap.Stream.Items {
			if _, ok := i.Event.(*ports.TextMessageContentEvent); ok {
				t.Fatal("the finished step's text is in the messages, not the items")
			}
		}
		mustEqual(t, snap.Stream.Offset != "", true, "an offset to read on from")
	})

	t.Run("a durable CUSTOM goes to the thread record; a plain one to the stream", func(t *testing.T) {
		h, streams := streamRuntime(t, scripted(step{calls: []call{{"c1", "go", `{}`}}}, step{text: "done"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{
			Name: "chat", Model: "gpt-4o",
			Tools: []agentenkit.Tool{agentenkit.AgentTool("go", "go", func(ctx context.Context, _ map[string]any, tc agentenkit.ToolContext) (string, error) {
				if _, err := tc.PublishEvent(ctx, "INVOICE_CREATED", map[string]any{"id": "inv_1"}, agentenkit.PublishOptions{Durable: true}); err != nil {
					return "", err
				}
				_, err := tc.PublishEvent(ctx, "PROGRESS", map[string]any{"pct": 50}, agentenkit.PublishOptions{})
				return "ok", err
			})},
		})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
		h.handleNext(t)
		stored, _ := h.storage.Events().ListSince(h.ctx, ran.ThreadID, -1, agentenkit.StorageContext{})
		var record []string
		for _, e := range stored {
			record = append(record, e.Type)
		}
		mustStrings(t, record, []string{"RUN_STARTED", "INVOICE_CREATED", "RUN_ENDED"}, "only the durable one is kept")
		var custom []string
		for _, i := range streamItems(t, streams, ran.RunID+":1") {
			if c, ok := i.Event.(*ports.CustomEvent); ok {
				custom = append(custom, c.Name)
			}
		}
		// Each reaches a tab once: the durable one as its record entry, the
		// plain one on the stream.
		mustStrings(t, custom, []string{"PROGRESS"}, "only the plain one is on the stream")
	})
}
